---
title: "BMI_parafac"
output: html_document
date: "2025-02-21"
---

```{r setup}
library(tidyverse)
library(ggplot2)
library(ggpubr)
library(stringr)
library(parafac4microbiome)
library(CMTFtoolbox)
```

```{r processing}
# Load subject info
infant_anthropometrics = read.csv("./Data/BMI/infant_anthropometrics.csv") %>% as_tibble()

# Faeces subject info
sampleInfo = read.csv("./Data/BMI/faeces_sampleMeta.csv", header=FALSE, sep=" ") %>% as_tibble()
colnames(sampleInfo) = c("Sample", "RCID", "BMI", "BMI.group", "Days", "Gestational.age", "C.section", "AB.infant", "AB.mother", "Secretor", "Lewis", "subject")
sampleInfo = sampleInfo %>% left_join(infant_anthropometrics %>% select(RCID, whz.6m))
subjectMeta_faeces = sampleInfo %>% select(subject, BMI, BMI.group, C.section, Secretor, Lewis, whz.6m) %>% unique() %>% mutate(subject=as.character(subject)) %>% arrange(subject)
subjectMeta_faeces = subjectMeta_faeces[-90,]

# Milk subject info
sampleInfo = read.csv("./Data/BMI/milk_sampleMeta.csv", header=FALSE, sep=" ") %>% as_tibble()
colnames(sampleInfo) = c("Sample", "RCID", "BMI", "BMI.group", "Days", "Gestational.age", "C.section", "AB.infant", "AB.mother", "Secretor", "Lewis", "subject")
subjectMeta = sampleInfo %>% select(subject, BMI, BMI.group, C.section, Secretor, Lewis) %>% unique() %>% mutate(subject=as.character(subject)) %>% arrange(subject)
subjectMeta_milk = subjectMeta %>% left_join(subjectMeta_faeces)

# Milk metab subject info
sampleInfo = read.csv("./Data/BMI/milkMetab_sampleMeta.csv", header=FALSE, sep=" ") %>% as_tibble()
colnames(sampleInfo) = c("RCID", "BMI", "BMI.group", "Days", "Gestational.age", "C.section", "AB.infant", "AB.mother", "Secretor", "Lewis", "subject")
subjectMeta = sampleInfo %>% select(subject, BMI, BMI.group, C.section, Secretor, Lewis) %>% unique() %>% mutate(subject=as.character(subject)) %>% arrange(subject)
subjectMeta_milkMetab = subjectMeta %>% left_join(subjectMeta_faeces %>% select(subject, whz.6m))

# Identify shared subjects
sharedSubjects = intersect(intersect(subjectMeta_faeces$subject, subjectMeta_milk$subject), subjectMeta_milkMetab$subject)
homogenized_subjectMeta_bmi = subjectMeta_faeces %>% filter(subject %in% sharedSubjects) %>% arrange(subject)
homogenized_subjectMeta_whz = homogenized_subjectMeta_bmi %>% filter(whz.6m != "NA")

# Faecal microbiome
df = read.csv("./Data/BMI/faecesCounts.csv", header=FALSE, sep=" ") %>% as_tibble()
taxonomy = read.csv("./Data/BMI/newTaxonomy_faeces.csv", header=FALSE, sep=" ") %>% as_tibble()
sampleInfo = read.csv("./Data/BMI/faeces_sampleMeta.csv", header=FALSE, sep=" ") %>% as_tibble()
colnames(sampleInfo) = c("Sample", "RCID", "BMI", "BMI.group", "Days", "Gestational.age", "C.section", "AB.infant", "AB.mother", "Secretor", "Lewis", "subject")

subjectMeta = sampleInfo %>% select(subject, BMI, BMI.group, C.section, Secretor, Lewis) %>% unique() %>% mutate(subject=as.character(subject)) %>% arrange(subject)

# Filter taxa to taxa with at least 1 non-zero value
featureMask = colSums(df) > 0
df = df[,featureMask]
taxonomy = taxonomy[featureMask,]

# Filter based on sparsity
threshold = 0.75

sparsity = colSums(df==0) / nrow(df)

featureSelection = sparsity <= threshold
taxonomy_filtered = taxonomy[featureSelection,]
df_filtered = df[,featureSelection]

# CLR
df_clr = t(apply(df_filtered+1, 1, function(x){log(x / compositions::geometricmean(x))})) %>% as_tibble()

# Make into cube
I = length(unique(sampleInfo$subject))
J = ncol(df_clr)
K = length(unique(sampleInfo$Days))
X = array(0L, c(I,J,K))
timepoints = sampleInfo %>% arrange(Days) %>% select(Days) %>% unique() %>% pull()

for(k in 1:K){
  Day = timepoints[k]
  X[,,k] = cbind(df_clr, sampleInfo) %>% as_tibble() %>% mutate(subject=as.character(subject)) %>% filter(Days == Day) %>% select(c(colnames(df_clr),subject)) %>% right_join(subjectMeta) %>% arrange(subject) %>% select(-colnames(subjectMeta)) %>% as.matrix()
}

# Mask based on shared subjects for BMI and WHZ
X_bmi = X[subjectMeta$subject %in% homogenized_subjectMeta_bmi$subject,,]
X_whz = X[subjectMeta$subject %in% homogenized_subjectMeta_whz$subject,,]

# Center and scale
X_bmi_cnt = parafac4microbiome::multiwayCenter(X_bmi, mode=1)
X_bmi_cnt_scl = parafac4microbiome::multiwayScale(X_bmi_cnt, mode=2)

X_whz_cnt = parafac4microbiome::multiwayCenter(X_whz, mode=1)
X_whz_cnt_scl = parafac4microbiome::multiwayScale(X_whz_cnt, mode=2)

# Save
faeces_df_bmi = X_bmi_cnt_scl
faeces_df_whz = X_whz_cnt_scl
faeces_taxonomy = taxonomy_filtered
faeces_timepoints = timepoints

# Milk microbiome
df = read.csv("./Data/BMI/milkCounts.csv", header=FALSE, sep=" ") %>% as_tibble()
taxonomy = read.csv("./Data/BMI/newTaxonomy_milk.csv", header=FALSE, sep=" ") %>% as_tibble()
sampleInfo = read.csv("./Data/BMI/milk_sampleMeta.csv", header=FALSE, sep=" ") %>% as_tibble()
colnames(sampleInfo) = c("Sample", "RCID", "BMI", "BMI.group", "Days", "Gestational.age", "C.section", "AB.infant", "AB.mother", "Secretor", "Lewis", "subject")

# Make subject metadata
subjectMeta = sampleInfo %>% select(subject, BMI, BMI.group, C.section, Secretor, Lewis) %>% unique() %>% mutate(subject=as.character(subject)) %>% arrange(subject)

# Filter taxa to taxa with at least 1 non-zero value
featureMask = colSums(df) > 0
df = df[,featureMask]
taxonomy = taxonomy[featureMask,]

# Filter based on sparsity
threshold = 0.85

sparsity = colSums(df==0) / nrow(df)

featureSelection = sparsity <= threshold
taxonomy_filtered = taxonomy[featureSelection,]
df_filtered = df[,featureSelection]

# CLR
df_clr = t(apply(df_filtered+1, 1, function(x){log(x / compositions::geometricmean(x))})) %>% as_tibble()

# Make into cube
I = length(unique(sampleInfo$subject))
J = ncol(df_clr)
K = length(unique(sampleInfo$Days))
X = array(0L, c(I,J,K))
timepoints = sampleInfo %>% arrange(Days) %>% select(Days) %>% unique() %>% pull()

for(k in 1:K){
  Day = timepoints[k]
  X[,,k] = cbind(df_clr, sampleInfo) %>% as_tibble() %>% mutate(subject=as.character(subject)) %>% filter(Days == Day) %>% select(c(colnames(df_clr),subject)) %>% right_join(subjectMeta) %>% arrange(subject) %>% select(-colnames(subjectMeta)) %>% as.matrix()
}

# Mask based on shared subjects for BMI and WHZ
X_bmi = X[subjectMeta$subject %in% homogenized_subjectMeta_bmi$subject,,]
X_whz = X[subjectMeta$subject %in% homogenized_subjectMeta_whz$subject,,]

# Center and scale
X_bmi_cnt = parafac4microbiome::multiwayCenter(X_bmi, mode=1)
X_bmi_cnt_scl = parafac4microbiome::multiwayScale(X_bmi_cnt, mode=2)

X_whz_cnt = parafac4microbiome::multiwayCenter(X_whz, mode=1)
X_whz_cnt_scl = parafac4microbiome::multiwayScale(X_whz_cnt, mode=2)

milk_df_bmi = X_bmi_cnt_scl
milk_df_whz = X_whz_cnt_scl
milk_taxonomy = taxonomy_filtered
milk_timepoints = timepoints

# Milk metabolomics
df = read.csv("./Data/BMI/milkMetabNumeric.csv", header=FALSE, sep=" ") %>% as_tibble()
taxonomy = read.csv("./Data/BMI/milk_metab_CAS_numbers.csv", header=TRUE, sep=",") %>% as_tibble()
sampleInfo = read.csv("./Data/BMI/milkMetab_sampleMeta.csv", header=FALSE, sep=" ") %>% as_tibble()
colnames(sampleInfo) = c("RCID", "BMI", "BMI.group", "Days", "Gestational.age", "C.section", "AB.infant", "AB.mother", "Secretor", "Lewis", "subject")
metaboliteCategories = read.csv("./Data/BMI/Metabolite_categories.txt", sep="\t") %>% as_tibble()
metaboliteCategories = metaboliteCategories %>% mutate(Metabolite = vctrs::vec_as_names(Metabolite, repair="universal_quiet"))

# Make subject metadata
subjectMeta = sampleInfo %>% select(subject, BMI, BMI.group, C.section, Secretor, Lewis) %>% unique() %>% mutate(subject=as.character(subject)) %>% arrange(subject)

# Remove duplicates
drop = c(61,81,146)
subjectMeta = subjectMeta %>% mutate(index=1:nrow(.)) %>% filter(!index %in% drop) %>% select(-index) %>% arrange(subject)

# Log transform
df_log = log(df)

# Make into cube
I = length(unique(sampleInfo$subject))
J = ncol(df_log)
K = length(unique(sampleInfo$Days))
X = array(0L, c(I,J,K))
timepoints = sampleInfo %>% arrange(Days) %>% select(Days) %>% unique() %>% pull()

for(k in 1:K){
  Day = timepoints[k]
  X[,,k] = cbind(df_log, sampleInfo) %>% as_tibble() %>% mutate(subject=as.character(subject)) %>% filter(Days == Day) %>% select(c(colnames(df_log),subject)) %>% right_join(subjectMeta) %>% arrange(subject) %>% select(-colnames(subjectMeta)) %>% as.matrix()
}

# Mask based on shared subjects for BMI and WHZ
X_bmi = X[subjectMeta$subject %in% homogenized_subjectMeta_bmi$subject,,]
X_whz = X[subjectMeta$subject %in% homogenized_subjectMeta_whz$subject,,]

# Center and scale
X_bmi_cnt = parafac4microbiome::multiwayCenter(X_bmi, mode=1)
X_bmi_cnt_scl = parafac4microbiome::multiwayScale(X_bmi_cnt, mode=2)

X_whz_cnt = parafac4microbiome::multiwayCenter(X_whz, mode=1)
X_whz_cnt_scl = parafac4microbiome::multiwayScale(X_whz_cnt, mode=2)

milkMetab_df_bmi = X_bmi_cnt_scl
milkMetab_df_whz = X_whz_cnt_scl
milkMetab_taxonomy = taxonomy
milkMetab_timepoints = timepoints

# Fix milkMetab feature annotations
df = milkMetab_taxonomy %>% mutate(Metabolite = make.names(X), CAS.Registry = make.names(CAS.Registry)) %>% left_join(metaboliteCategories)

# Fix mismatches by hand
df[df$Metabolite == "X2.Aminobutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X2.Hydroxybutyrate","Class"] = "Amino acids and derivatives"
df[df$Metabolite == "X2.Oxoglutarate","Class"] = "Energy related"
df[df$Metabolite == "X3.Fucosyllactose","Class"] = "Oligosaccharides"
df[df$Metabolite == "X3SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "X6SL.partial","Class"] = "Oligosaccharides"
df[df$Metabolite == "Methionine","Class"] = "Amino acids and derivatives"
df[70,"Metabolite"] = "tau.Methylhistidine" # Fix non-ascii tau character

milkMetab_featureMeta = df %>% select(-X)
```

```{r parafac of HM metabolomics}
# result = parafac4microbiome::assessModelQuality(milkMetab_df_bmi, numRepetitions=10, numCores=parallel::detectCores())
# result = parafac4microbiome::assessModelStability(milkMetab_df_bmi, numFolds=10, numCores=parallel::detectCores())
milkMetab_parafac = parafac4microbiome::parafac(milkMetab_df_bmi, nfac=1, nstart=100)
```

```{r parafac of HM microbiome}
# result = parafac4microbiome::assessModelQuality(milk_df_bmi, numRepetitions=10, numCores=parallel::detectCores())
milk_parafac = parafac4microbiome::parafac(milk_df_bmi, nfac=1, nstart=100)
```

```{r parafac of infant faecal microbiome}
# result = parafac4microbiome::assessModelQuality(faeces_df_bmi, numRepetitions=10, numCores=parallel::detectCores())
# result = parafac4microbiome::assessModelQuality(faeces_df_bmi, numRepetitions=10, numCores=parallel::detectCores())
faeces_parafac = parafac4microbiome::parafac(faeces_df_bmi, nfac=1, nstart=100)
```

```{r plot HM metabolomics model}
prepMode1 = homogenized_subjectMeta_bmi %>% mutate(Component_1 = milkMetab_parafac$Fac[[1]]) %>% arrange(Secretor,subject) %>% mutate(index=1:nrow(.))
prepMode2 = milkMetab_featureMeta %>% mutate(Component_1 = milkMetab_parafac$Fac[[2]]) %>% arrange(Class, Metabolite) %>% mutate(index=1:nrow(.))

a = prepMode1 %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(Secretor))) + geom_bar(stat="identity")
b = prepMode2 %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(Class))) + geom_bar(stat="identity")
c = milkMetab_parafac$Fac[[3]] %>% as_tibble() %>% mutate(timepoint = c(3,30,60,90)) %>% ggplot(aes(x=timepoint,y=V1)) + geom_line() + geom_point()

ggarrange(a,b,c,nrow=1,ncol=3)
```

```{r plot HM microbiome model}
prepMode1 = homogenized_subjectMeta_bmi %>% mutate(Component_1 = milk_parafac$Fac[[1]]) %>% arrange(BMI.group,subject) %>% mutate(index=1:nrow(.))
prepMode2 = milk_taxonomy %>% mutate(Component_1 = milk_parafac$Fac[[2]]) %>% arrange(V3, V1) %>% mutate(index=1:nrow(.))

a = prepMode1 %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(BMI.group))) + geom_bar(stat="identity")
b = prepMode2 %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(V3))) + geom_bar(stat="identity")
c = milk_parafac$Fac[[3]] %>% as_tibble() %>% mutate(timepoint = c(3,30,60,90)) %>% ggplot(aes(x=timepoint,y=V1)) + geom_line() + geom_point()

ggarrange(a,b,c,nrow=1,ncol=3)
```
```{r plot faecal microbiome model}
prepMode1 = homogenized_subjectMeta_bmi %>% mutate(Component_1 = faeces_parafac$Fac[[1]]) %>% arrange(BMI.group,subject) %>% mutate(index=1:nrow(.))
prepMode2 = faeces_taxonomy %>% mutate(Component_1 = faeces_parafac$Fac[[2]]) %>% arrange(V3, V1) %>% mutate(index=1:nrow(.))

a = prepMode1 %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(BMI.group))) + geom_bar(stat="identity")
b = prepMode2 %>% ggplot(aes(x=index,y=Component_1,fill=as.factor(V3))) + geom_bar(stat="identity")
c = faeces_parafac$Fac[[3]] %>% as_tibble() %>% mutate(timepoint = c(30,60,90)) %>% ggplot(aes(x=timepoint,y=V1)) + geom_line() + geom_point()

ggarrange(a,b,c,nrow=1,ncol=3)
```

```{r interpret loadings}
testAssociationsPerComponent = function(model){
  numComponents = ncol(model$Fac[[1]])
  
  result = matrix(NA, nrow=numComponents, ncol=5)
  
  for(i in 1:numComponents){
    result[i,1] = cor.test(model$Fac[[1]][,i], Jakobsen2025$subjectMeta_BMI$BMI, method="spearman")$p.value
    result[i,2] = cor.test(model$Fac[[1]][,i], Jakobsen2025$subjectMeta_BMI$whz.6m, method="spearman")$p.value
    
    mask = !is.na(Jakobsen2025$subjectMeta_BMI$C.section)
    temp = model$Fac[[1]][mask,i]
    tempMeta = Jakobsen2025$subjectMeta_BMI[mask,]
    mask1 = tempMeta$C.section == 1
    mask2 = tempMeta$C.section == 2
    
    result[i,3] = wilcox.test(temp[mask1], temp[mask2])$p.value
    
    mask1 = Jakobsen2025$subjectMeta_BMI$Secretor == 0
    mask2 = Jakobsen2025$subjectMeta_BMI$Secretor == 1
    result[i,4] = wilcox.test(model$Fac[[1]][mask1,i], model$Fac[[1]][mask2,i])$p.value
    
    mask1 = Jakobsen2025$subjectMeta_BMI$Lewis == 0
    mask2 = Jakobsen2025$subjectMeta_BMI$Lewis == 1
    result[i,5] = wilcox.test(model$Fac[[1]][mask1,i], model$Fac[[1]][mask2,i])$p.value
  }
  
  return(result)
}

result1 = testAssociationsPerComponent(faeces_parafac)
result2 = testAssociationsPerComponent(milk_parafac)
result3 = testAssociationsPerComponent(milkMetab_parafac)

df = rbind(result1, result2, result3)
corrected_df = signif(matrix(p.adjust(df, "BH"), dim(df)), 2)
```
